### ISSUE TITLE
How to complete handshake?

### ISSUE THREAD

**Issue by stepancheg** [ctz/rustls#77]
Could you please provide a simple example how to complete a handshake with `rustls`?

I'm trying this:

```
        while self.session.is_handshaking() {
            if self.session.wants_write() {
                self.session.write_tls(&mut self.stream)?;
            }
            if self.session.wants_read() {
                self.session.read_tls(&mut self.stream)?;
            }
            self.session.process_new_packets().expect("TODO");
        }
```

It hangs I have now idea why.

**Comment by stepancheg** [comment]
And this hangs too:

```
        while self.session.is_handshaking() {
            if self.session.is_handshaking() && self.session.wants_write() {
                self.session.write_tls(&mut self.stream)?;
            }
            if self.session.is_handshaking() && self.session.wants_read() {
                self.session.read_tls(&mut self.stream)?;
                self.session.process_new_packets().expect("TODO");
            }
        }
```

`stream` is a sync `TcpStream` connected to `google.com:443`

**Comment by stepancheg** [comment]
Debug output: https://gist.github.com/stepancheg/d5b4bd4ad2aa45302afc88fed220603e

**Comment by stepancheg** [comment]
OK, I've found a solution:

```
        while self.session.is_handshaking() {
            if self.session.is_handshaking() && self.session.wants_write() {
                while self.session.write_tls(&mut self.stream)? > 0 {
                };
            }
            if self.session.is_handshaking() && self.session.wants_read() {
                self.session.read_tls(&mut self.stream)?;
                self.session.process_new_packets().expect("TODO");
            }
        }
```

So I believe there's a bug in rustls: `wants_read` returns true, but it hangs at `read_tls` because write buffer is not flushed.

**Comment by ctz** [comment]
In practice, a TLS session always wants to receive data because the peer can send alerts at any time and the session needs to be prepared to receive them. That's why you experience `wants_read` returning true at this time.

It is correct to continue calling `write_tls` while `wants_write` yields true. `write_tls` does not coalesce writes at the moment for copy-efficiency reasons (coalescing would involve extra allocations and copies, because `std::io::Write` doesn't support scattered writes; standard writev support would be excellent here).

Sorry for the delay in explaining this; I was on holiday this week.

I'm going to close this bug, and I assume your difficulties here will be best addressed in #75 -- the implementation of that will effectively be the code you wanted to write here.

**Comment by stepancheg** [comment]
If that `TlsStream` object (hiding writes, reads and buffers) will have an operation like `complete_handshake` which will work with both blocking and non-blocking sockets, that will be perfect for me.
